File: src/ui/cursor.py
"""Always-on pinch cursor overlay.

One cursor per tracked hand, drawn at the pinch point the detector actually
uses (thumb-tip-anchored, One-Euro-smoothed, latency-compensated — see
detection/gestures.py). A ring fills with the continuous pinch progress
(0 at the release threshold, 1 at the close threshold — Meta-style pinch
strength as a UI signal) and a short expanding flash confirms each
registered click. This closes the feedback loop that made pinching feel
rough: the user steers the dot they can see, and every state the machine
passes through is visible the frame it happens.
"""

import time

import cv2

from detection import gestures

# Palette (BGR) — visuals stay module-local (hints.py / slingshot precedent).
COL_IDLE = (180, 180, 180)     # neutral ring, hand open
COL_CLOSING = (60, 200, 255)   # amber while the fingers close
COL_HELD = (140, 255, 150)     # green while the pinch is held
COL_FLASH = (255, 255, 255)    # click confirmation ring
COL_UNDER = (25, 25, 25)       # dark under-stroke for video readability

RING_R = 14        # progress ring radius (px)
DOT_R = 4          # centre dot radius (px)
FLASH_FRAMES = 6   # duration of the click flash
FLASH_GROW_PX = 18 # how far the flash ring expands
STALE_S = 0.2      # hide machines not updated this recently (grace ghosts)


class PinchCursor:
    """Draws the per-hand pinch cursors. Reads the gesture snapshot only —
    construct once and call ``draw(frame)`` after the scene, before the
    onboarding overlays."""

    def __init__(self, frame_w, frame_h):
        self.w = frame_w
        self.h = frame_h
        self._flash = {}   # hand_id -> frames left of the click flash

    def draw(self, frame):
        now = time.monotonic()
        live = set()
        for hid, m in gestures.pinch_infos():
            # A machine inside the tracking-grace window but without fresh
            # data would freeze a ghost cursor on screen — skip it.
            if now - m.last_seen > STALE_S:
                continue
            live.add(hid)
            if m.pinching:
                self._flash[hid] = FLASH_FRAMES

            x, y = int(m.cursor[0]), int(m.cursor[1])
            if not (0 <= x < self.w and 0 <= y < self.h):
                continue
            progress = m.progress
            if m.closed:
                color = COL_HELD
            elif progress > 0.0:
                color = COL_CLOSING
            else:
                color = COL_IDLE

            # Faint base ring, always visible (this IS the cursor).
            cv2.circle(frame, (x, y), RING_R, COL_UNDER, 4, cv2.LINE_AA)
            cv2.circle(frame, (x, y), RING_R, COL_IDLE, 1, cv2.LINE_AA)
            # Progress arc fills clockwise from 12 o'clock as fingers close.
            if progress > 0.0:
                end = int(-90 + 360 * progress)
                cv2.ellipse(frame, (x, y), (RING_R, RING_R), 0, -90, end,
                            COL_UNDER, 5, cv2.LINE_AA)
                cv2.ellipse(frame, (x, y), (RING_R, RING_R), 0, -90, end,
                            color, 3, cv2.LINE_AA)
            # Centre dot; slightly larger and colored while held.
            r = DOT_R + (2 if m.closed else 0)
            cv2.circle(frame, (x, y), r + 2, COL_UNDER, -1, cv2.LINE_AA)
            cv2.circle(frame, (x, y), r, color, -1, cv2.LINE_AA)

            # Click flash: an expanding, thinning ring for a few frames.
            f = self._flash.get(hid, 0)
            if f > 0:
                t = f / FLASH_FRAMES
                fr = int(RING_R + (1.0 - t) * FLASH_GROW_PX)
                cv2.circle(frame, (x, y), fr, COL_FLASH,
                           max(1, int(3 * t)), cv2.LINE_AA)
                self._flash[hid] = f - 1

        # Drop flash counters of departed hands.
        self._flash = {h: v for h, v in self._flash.items()
                       if h in live and v > 0}

File: web/src/overlay/cursor.ts
import type { HandState } from "../state/types";

/**
 * Per-hand pinch cursor — the web port of `src/ui/cursor.py`.
 *
 * Same affordance, redrawn in the design system: a ring whose arc fills
 * clockwise from 12 o'clock with the pinch progress (amber while closing,
 * green while held), a centre dot, and an expanding flash confirming each
 * registered click. The onboarding hand demo shows this exact ring, so
 * users recognize it on their own hand.
 */

const RING_R = 16;
const DOT_R = 4;
const STALE_MS = 200; // matches cursor.py STALE_S
const FLASH_MS = 260;
const FLASH_GROW = 22;

const UNDER = "rgba(8,10,15,0.75)";
const IDLE = "rgba(237,241,247,0.65)";
const AMBER = "#ffb84d";
const GREEN = "#63e6a4";
const FLASH = "#ffffff";

export function drawCursors(
  ctx: CanvasRenderingContext2D,
  hands: HandState[],
  flashes: Map<string, number>,
  now: number,
  frameW: number,
  frameH: number,
) {
  const live = new Set<string>();
  for (const hand of hands) {
    if (hand.seen_ms > STALE_MS) continue; // grace-window ghost
    live.add(hand.id);

    const [x, y] = hand.cursor;
    if (x < 0 || y < 0 || x >= frameW || y >= frameH) continue;

    const color =
      hand.held ? GREEN : hand.progress > 0 ? AMBER : IDLE;

    // Base ring: dark under-stroke for readability on any video content.
    ctx.lineCap = "round";
    ctx.strokeStyle = UNDER;
    ctx.lineWidth = 4.5;
    ctx.beginPath();
    ctx.arc(x, y, RING_R, 0, Math.PI * 2);
    ctx.stroke();
    ctx.strokeStyle = IDLE;
    ctx.lineWidth = 1.5;
    ctx.beginPath();
    ctx.arc(x, y, RING_R, 0, Math.PI * 2);
    ctx.stroke();

    // Progress arc from 12 o'clock.
    if (hand.progress > 0) {
      const start = -Math.PI / 2;
      const end = start + Math.PI * 2 * hand.progress;
      ctx.strokeStyle = UNDER;
      ctx.lineWidth = 5.5;
      ctx.beginPath();
      ctx.arc(x, y, RING_R, start, end);
      ctx.stroke();
      ctx.strokeStyle = color;
      ctx.lineWidth = 3;
      ctx.beginPath();
      ctx.arc(x, y, RING_R, start, end);
      ctx.stroke();
    }

    // Centre dot, larger while held.
    const r = DOT_R + (hand.held ? 2 : 0);
    ctx.fillStyle = UNDER;
    ctx.beginPath();
    ctx.arc(x, y, r + 2, 0, Math.PI * 2);
    ctx.fill();
    ctx.fillStyle = color;
    ctx.beginPath();
    ctx.arc(x, y, r, 0, Math.PI * 2);
    ctx.fill();

    // Click flash: expanding, thinning, fading ring.
    const started = flashes.get(hand.id);
    if (started !== undefined) {
      const p = (now - started) / FLASH_MS;
      if (p >= 1) {
        flashes.delete(hand.id);
      } else {
        ctx.strokeStyle = FLASH;
        ctx.globalAlpha = 1 - p;
        ctx.lineWidth = 3 * (1 - p) + 0.5;
        ctx.beginPath();
        ctx.arc(x, y, RING_R + p * FLASH_GROW, 0, Math.PI * 2);
        ctx.stroke();
        ctx.globalAlpha = 1;
      }
    }
  }

  // Drop flash timers of hands that left the frame.
  for (const id of flashes.keys()) {
    if (!live.has(id)) flashes.delete(id);
  }
}
